### PR TITLE
Eamon/line items

### PR DESCRIPTION
**Pull request by eito-fis**

### CODE REVIEW

File: tests/test_sequence_labeling_encoder.py [hunk 1]
```diff
@@ -65,3 +67,24 @@ def test_sequence_label_encoder_half():
...
+    out = EncodedOutput(
+        token_ids=np.array([   0, 9583,  139,   40,  249, 8875,    2]), 
+        tokens=np.array(['0', 'five', ' per', 'cent', ' (', '5', '%)', '2'], dtype='<U21'), 
+        token_ends=np.array([-1,  4, 8, 12, 14, 15, 17, -1]), 
+        token_starts=np.array([-1,  0,  5, 8, 13, 14, 15, -1]), 
+        useful_start=0, 
+        useful_end=512
+    )
+    label_arr = encoder.transform(out, labels)
+    assert label_arr == [0, 1, 1, 2, 1, 2, 1, 0]
```
Comment (madisonmay) [inline tests/test_sequence_labeling_encoder.py]: I'm struggling to see how this is the expected output for BIO tagging.  Assuming 1 represents "B", where is the third 1 coming from?

File: tests/test_group_target_encoder.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        token_starts=np.array([-1,  0,  5, 8, 13, 14, 15, -1]), 
+        useful_start=0, 
+        useful_end=512
+    )
+    label_arr = encoder.transform(out, label)
+    print(label_arr)
+    input()
+    assert label_arr == [
+        [[0, 0, 0, 0, 0, 0, 0, 0],
+         [0, 0, 0, 0, 1, 1, 1, 0],
```
Comment (madisonmay) [inline tests/test_group_target_encoder.py:199]: I wouldn't expect the third 1 here -- shouldn't the token "five" be connected to two other tokens only?

File: tests/test_group_target_encoder.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        tokens=np.array(['0', 'five', ' per', 'cent', ' (', '5', '%)', '2'], dtype='<U21'), 
+        token_ends=np.array([-1,  4, 8, 12, 14, 15, 17, -1]), 
+        token_starts=np.array([-1,  0,  5, 8, 13, 14, 15, -1]), 
+        useful_start=0, 
+        useful_end=512
+    )
+    label_arr = encoder.transform(out, label)
+    print(label_arr)
+    input()
+    assert label_arr == [
```
Comment (madisonmay) [inline tests/test_group_target_encoder.py:197]: Undirected graph where edge=1 indicates membership in same group

File: tests/test_group_target_encoder.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        [[0, 0, 0, 0, 0, 0, 0, 0],
```
Comment (madisonmay) [inline tests/test_group_target_encoder.py:209]: Having trouble deciphering what intended behavior is here. Would you mind briefly explaining what the arrays below represent?

File: finetune/encoding/target_encoders.py [hunk 5]
```diff
@@ -229,21 +249,338 @@ def transform(self, out, labels):
...
-                    if label["label"] not in self.lookup:
+                    if current_label not in self.lookup:
                         LOGGER.warning(
                             "Attempting to encode unknown labels : {}, ignoring for now but this will likely not "
-                            "result in desirable behaviour. Available labels are {}".format(label["label"], self.lookup.keys())
+                            "result in desirable behaviour. Available labels are {}".format(current_label, self.lookup.keys())
                         )
                     else:
-                        labels_out[i] = self.lookup[label["label"]]
+                        labels_out[i] = self.lookup[current_label]
```
Comment (madisonmay) [inline finetune/encoding/target_encoders.py:311]: Bit uncomfortable that output space is 2 x 3 x n_classes when bio_tagging and group_tagging is used but I guess if the same classes nearly always begin the group there's no issue

File: tests/test_group_target_encoder.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        token_ends=np.array([-1,  4, 8, 12, 14, 15, 17, -1]), 
+        token_starts=np.array([-1,  0,  5, 8, 13, 14, 15, -1]), 
+        useful_start=0, 
+        useful_end=512
+    )
+    label_arr = encoder.transform(out, label)
+    print(label_arr)
+    print(encoder.classes_)
+    input()
+    assert label_arr == [[0, 0, 1, 1, 0, 0, 0, 0], [0, 0, 4, 0, 5, 6, 0, 0]]
```
Comment (madisonmay) [inline tests/test_group_target_encoder.py:137]: - Binary indicator of group start
- Index of next token in chain

File: tests/test_group_target_encoder.py (hunk 4 above)
Comment (madisonmay) [reply tests/test_group_target_encoder.py:209]: Based on my understanding of the code, if arr[i][j] = 1 then the token at index i is part of the same label as j. Is this intended to be part of the same label class, or part of the same label span?

File: tests/test_sequence_labeling_encoder.py (hunk 1 above)
Comment (eito-fis) [reply tests/test_sequence_labeling_encoder.py]: Test was broken, thanks for the catch.

File: finetune/target_models/sequence_labeling.py [hunk 7]
```diff
@@ -354,6 +357,15 @@ def predict(
 
     def _predict(
         self, zipped_data, per_token=False, return_negative_confidence=False, **kwargs
+    ):
+        predictions = list(self.process_long_sequence(zipped_data, **kwargs))
```
Comment (benleetownsend) [inline finetune/target_models/sequence_labeling.py]: This list conversion breaks handling predictions as an iterator

**Pull request closed without merging**